File: algorithms/divide-and-conquer/quick-sort.ts
/**
 * Quick Sort
 *
 * Auswählen eines Pivot-Elements und sortieren der restlichen Elemente in ein "left" und ein "right" Array.
 *
 * Komplexität: O(n^2)
 */
const quickSort = (items: number[]) => {
  if (items.length <= 1) {
    return items;
  }
  const pivot = items[0];
  const left = [];
  const right = [];
  for (let i = 1; i < items.length; i++) {
    if (items[i] < pivot) {
      left.push(items[i]);
    } else {
      right.push(items[i]);
    }
  }
  return [...quickSort(left), pivot, ...quickSort(right)];
};

const result = quickSort([4, 8, 7, 2, 11, 1, 3]);
console.log(`Output: ${result}`);
export {};

File: algorithms/divide-and-conquer/quick-sort.py
"""
Quick Sort

Auswählen eines Pivot-Elements und sortieren der restlichen Elemente in ein "left" und ein "right" Array.

Komplexität: O(n^2)
"""


def quickSort(values: list) -> list:
    if len(values) < 2:
        return values
    pivot = values[0]
    right = []
    left = []
    for value in values[1:]:
        if value > pivot:
            right.append(value)
        else:
            left.append(value)
    return quickSort(left) + [pivot] + quickSort(right)


if __name__ == "__main__":
    print(quickSort([4, 8, 7, 2, 11, 1, 3]))
